astropy/coordinates/baseframe.py

```diff
@@ -709,8 +856,17 @@ def transform_to(self, new_frame):
         if self._data is None:
             raise ValueError('Cannot transform a frame with no data')
 
+        if (getattr(self.data, 'differentials', None) and
+           hasattr(self, 'obstime') and hasattr(new_frame, 'obstime') and
+           np.all(self.obstime == new_frame.obstime)):
```

Comment:
Possibly, it should be that if `obstime` is not given on a frame, it is assumed to be J2000 for this purposes.  Of course, on `SkyCoord`, this would just work, since there the obstime is propagated.